contracts/Joint.sol

```diff
@@ -159,85 +185,302 @@ contract Joint {
         }
 
         Joint(newJoint).initialize(
-            _governance,
-            _keeper,
-            _strategist,
-            _tokenA,
-            _tokenB,
-            _router
+            _providerA,
+            _providerB,
+            _router,
+            _weth,
+            _masterchef,
+            _reward,
+            _pid
         );
 
         emit Cloned(newJoint);
     }
 
-    function name() external view returns (string memory) {
-        string memory ab =
-            string(
-                abi.encodePacked(
-                    IERC20Extended(address(tokenA)).symbol(),
-                    IERC20Extended(address(tokenB)).symbol()
-                )
-            );
-
-        return string(abi.encodePacked("JointOf", ab));
-    }
+    function name() external view virtual returns (string memory) {}
+
+    function prepareReturn(bool returnFunds) external onlyProviders {
+        // If we have previously invested funds, let's distrubute PnL equally in
+        // each token's own terms
+        if (investedA != 0 && investedB != 0) {
+            // Track starting amount in case reward is one of LP tokens
+            uint256 startingRewardBal = balanceOfReward();
+
+            if (balanceOfStake() != 0) {
+                getReward();
+            }
+
+            uint256 rewardAmount = balanceOfReward().sub(startingRewardBal);
+
+            // Liquidate will also claim rewards
+            (uint256 currentA, uint256 currentB) = _liquidatePosition();
```

Comment:
you should take into account balanceOfA and balanceOfB in currentA and currentB

they might not be exactly the amount you just withdrew from DEX in case of airdrops or, most probably, not fully adding liquidity of both tokens like @tonkers-kuma correctly stated (as addLiquidity takes only the amount required to deposit both tokens in same proportion)